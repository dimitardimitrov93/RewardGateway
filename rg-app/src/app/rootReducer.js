import { combineReducers } from 'redux';
import employeeReducer from './employee/employeeReducer';
import { persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

const persistConfig = {
    key: 'root',
    storage,
    whitelist: ['employee'],
}

const rootReducer = combineReducers({
    employee: employeeReducer,
})

export default persistReducer(persistConfig, rootReducer);