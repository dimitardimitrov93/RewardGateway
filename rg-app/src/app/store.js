import { createStore, applyMiddleware } from '@reduxjs/toolkit';
import { persistStore } from 'redux-persist';
import thunk from 'redux-thunk';
import rootReducer from './rootReducer';

export const store = createStore(rootReducer, applyMiddleware(thunk));
export const persistor = persistStore(store);

export default { store, persistor};
