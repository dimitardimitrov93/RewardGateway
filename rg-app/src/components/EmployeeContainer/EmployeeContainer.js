import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { fetchEmployees, updateEmployee, getAllEmployees } from '../../app/index'
import EmployeeCard from '../EmployeeCard/EmployeeCard';
import SearchComponent from '../SearchComponent/SearchComponent'
import { Container, Pagination } from '@mui/material'
import ArrowBackIcon from '@mui/icons-material/ArrowBack';
import Button from '@mui/material/Button';
import Masonry from 'react-masonry-css'

function EmployeeContainer({ employeeData, fetchEmployees, updateEmployee }) {

    const [employees, setEmployees] = useState(employeeData);
    const [searchQuery, setSearchQuery] = useState('');
    const [searchField, setSearchField] = useState(null);
    const [searchResults, setSearchResults] = useState([]);
    const [currentPage, setCurrentPage] = useState(1);
    const [employeesPerPage, setEmployeesPerPage] = useState(20);
    const indexOfLastEmployee = currentPage * employeesPerPage;
    const indexOfFirstEmployee = indexOfLastEmployee - employeesPerPage;
    const currentEmployees = employees.employees.slice(indexOfFirstEmployee, indexOfLastEmployee);
    const pageNumber = Math.ceil(employees.employees.length / employeesPerPage);

    const breakpoints = {
        default: 4,
        1600: 3,
        1100: 2,
        700: 1,
    }

    useEffect(() => {
        if (!employeeData.employees[0]) {
            fetchEmployees();
        }
    }, [])

    useEffect(() => {
        setEmployees(employeeData);
    }, [employeeData])

    function handleUpdateEmployee(sanitizedEmployeeInfo) {
        updateEmployee(sanitizedEmployeeInfo);

        if (!searchResults[0]) return;
        const newResults = searchResults.slice();
        let employeeIndex = null;

        for (let i = 0; i < newResults.length; i++) {
            if (newResults[i].uuid === sanitizedEmployeeInfo.uuid) {
                employeeIndex = i;
                break;
            }
        }
        newResults.splice(employeeIndex, 1, sanitizedEmployeeInfo);

        setSearchResults(newResults);
    }

    function onPageChange(e) {
        setCurrentPage(Number(e.target.textContent));
    }

    function onSearchChange(e) {
        if (!searchField) {
            setSearchField(e.target);
        } else {
            setSearchQuery(searchField.value.toLowerCase());
        }
    }

    function handleSearch(e) {
        e.preventDefault();

        if (!searchQuery) {
            setSearchResults([]);
            showNoResultsMessage();
            return;
        };

        const results = employeeData.employees.filter(employee => {
            if (employee['label']) {
                return employee['label'].toLowerCase().includes(searchQuery);
            }
        });

        if (results.length === 0) {
            showNoResultsMessage();
        }

        setSearchResults(results);
    }

    function clearSearchResults() {
        setSearchQuery('');
        setSearchResults([]);
        searchField.value = '';
        return;
    }

    function showNoResultsMessage() {
        document.getElementById('searchQueryResultMessage').style.display = "block";

        setTimeout(() => {
            document.getElementById('searchQueryResultMessage').style.display = "none";
        }, 3000);
    }

    return (
        <div>
            {employees.loading ?
                <h2>Loading</h2>
                : employees.error ? (
                    <h2>{employees.error}</h2>
                ) : (
                        <div>
                            <h2>Employee List</h2>
                            {employees && employees.employees &&
                                <>
                                    <SearchComponent onSearchChange={onSearchChange} handleSearch={handleSearch} />
                                    <Container data-testid="container-1">
                                        <h4 id="searchQueryResultMessage" style={{ display: "none", backgroundColor: "#8080807a", borderRadius: "1em", padding: "0.5em" }}>No results from this query.</h4>

                                        {searchResults.length > 0 &&
                                            <>
                                                <Button sx={{ marginBottom: "2em" }} variant="outlined" startIcon={<ArrowBackIcon />} onClick={clearSearchResults}>Back</Button>
                                                <Masonry
                                                    breakpointCols={breakpoints}
                                                    className="my-masonry-grid"
                                                    columnClassName="my-masonry-grid_column"
                                                >
                                                    {searchResults.map(employee => {
                                                        return <EmployeeCard
                                                            key={employee.uuid}
                                                            data-id={employee.uuid}
                                                            employee={employee}
                                                            handleUpdateEmployee={handleUpdateEmployee} />
                                                    })}
                                                </Masonry>
                                            </>
                                        }

                                        {searchResults.length === 0 &&
                                            <>
                                                <Masonry
                                                    breakpointCols={breakpoints}
                                                    className="my-masonry-grid"
                                                    columnClassName="my-masonry-grid_column"
                                                >
                                                    {currentEmployees.map(employee => {
                                                        return <EmployeeCard
                                                            key={employee.uuid}
                                                            data-id={employee.uuid}
                                                            employee={employee}
                                                            handleUpdateEmployee={handleUpdateEmployee} />
                                                    })}
                                                </Masonry>
                                                <Pagination
                                                    data-testid="pagination-1"
                                                    hideNextButton={true}
                                                    hidePrevButton={true}
                                                    color="primary"
                                                    sx={{
                                                        margin: "2em",
                                                        display: "flex",
                                                        justifyContent: "center"
                                                    }}
                                                    count={pageNumber}
                                                    onChange={onPageChange}
                                                />
                                            </>
                                        }

                                    </Container>
                                </>
                            }
                        </div>
                    )
            }
        </div>
    )
}

const mapStateToProps = (state) => {
    return {
        employeeData: state.employee,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        fetchEmployees: () => dispatch(fetchEmployees()),
        updateEmployee: (sanitizedEmployeeInfo) => dispatch(updateEmployee(sanitizedEmployeeInfo)),
        getAllEmployees: () => dispatch(getAllEmployees()),
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EmployeeContainer);
