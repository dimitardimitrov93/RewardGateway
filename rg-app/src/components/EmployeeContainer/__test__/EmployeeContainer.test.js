import React from 'react';
import ReactDOM from 'react-dom';
import EmployeeContainer from '../EmployeeContainer';
import { store, persistor } from '../../../app/store';
import { Provider } from 'react-redux';
import { render, screen, cleanup } from '@testing-library/react';
import { PersistGate } from 'redux-persist/integration/react'
import renderer from 'react-test-renderer';

it('renders EmployeeContainer without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render((<Provider store={store}>
        <PersistGate persistor={persistor}>
            <EmployeeContainer />
        </PersistGate>
    </Provider>), div);


    ReactDOM.unmountComponentAtNode(div);
});

it('renders Container component correctly', () => {
    render((
        <Provider store={store}>
            <PersistGate persistor={persistor}>
                <EmployeeContainer />
            </PersistGate>
        </Provider>
    ));
    const containerElement = screen.getByTestId('container-1');
    expect(containerElement).toBeInTheDocument();
});

it('renders Pagination component correctly', () => {
    render((
        <Provider store={store}>
            <PersistGate persistor={persistor}>
                <EmployeeContainer />
            </PersistGate>
        </Provider>
    ));
    const paginationElement = screen.getByTestId('pagination-1');
    expect(paginationElement).toBeInTheDocument();
});

test('matches snapshot', () => {
    const tree = renderer.create((
        <Provider store={store}>
            <PersistGate persistor={persistor}>
                <EmployeeContainer />
            </PersistGate>
        </Provider>
    )).toJSON();
    expect(tree).toMatchSnapshot();
})

afterEach(() => {
    cleanup();
});