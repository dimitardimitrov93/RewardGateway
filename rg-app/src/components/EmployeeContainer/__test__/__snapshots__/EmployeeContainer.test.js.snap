// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div>
    <h2>
      Employee List
    </h2>
    <form
      autoComplete="off"
      className="MuiBox-root css-y1rufk"
      data-testid="box-1"
      noValidate={true}
    >
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        data-testid="textField-1"
      >
        <label
          className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="outlined-basic"
          id="outlined-basic-label"
        >
          Search by label
        </label>
        <div
          className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id="outlined-basic"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-dhh0px"
            >
              <span>
                Search by label
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-39tiqt-MuiButtonBase-root-MuiButton-root"
        data-testid="button-1"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="SearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
    <div
      className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      data-testid="container-1"
    >
      <h4
        id="searchQueryResultMessage"
        style={
          Object {
            "backgroundColor": "#8080807a",
            "borderRadius": "1em",
            "display": "none",
            "padding": "0.5em",
          }
        }
      >
        No results from this query.
      </h4>
      <div
        className="my-masonry-grid"
      />
      <nav
        aria-label="pagination navigation"
        className="MuiPagination-root MuiPagination-text css-1r2vklm-MuiPagination-root"
        data-testid="pagination-1"
      >
        <ul
          className="MuiPagination-ul css-wjh20t-MuiPagination-ul"
        />
      </nav>
    </div>
  </div>
</div>
`;
